lib/internal/test_runner/runner.js

```diff
@@ -0,0 +1,159 @@
+    // .cjs, and .mjs files in the 'test' directory.
+    const isTestDir = name === 'test';
+    const { underTestDir } = options;
+    const entries = readdirSync(path);
+
+    if (isTestDir) {
+      options.underTestDir = true;
+    }
+
+    options.userSupplied = false;
+
+    for (let i = 0; i < entries.length; i++) {
+      processPath(join(path, entries[i]), testFiles, options);
+    }
+
+    options.underTestDir = underTestDir;
+  }
+}
+
+function createTestFileList() {
+  const cwd = process.cwd();
+  const hasUserSuppliedPaths = process.argv.length > 1;
+  const testPaths = hasUserSuppliedPaths ?
+    ArrayPrototypeSlice(process.argv, 1) : [cwd];
+  const testFiles = new SafeSet();
+
+  try {
+    for (let i = 0; i < testPaths.length; i++) {
+      const absolutePath = resolve(testPaths[i]);
+
+      processPath(absolutePath, testFiles, { userSupplied: true });
+    }
+  } catch (err) {
+    if (err?.code === 'ENOENT') {
+      console.error(`Could not find '${err.path}'`);
+      process.exit(1);
+    }
+
+    throw err;
+  }
+
+  return ArrayPrototypeSort(ArrayFrom(testFiles));
+}
+
+function filterExecArgv(arg) {
+  return !ArrayPrototypeIncludes(kFilterArgs, arg);
+}
+
+function runTestFile(path, root) {
+  const subtest = root.createSubtest(Test, path, async (t) => {
+    const args = ArrayPrototypeFilter(process.execArgv, filterExecArgv);
+    ArrayPrototypePush(args, path);
+
+    const child = spawn(process.execPath, args, { signal: t.signal, encoding: 'utf8' });
+    // TODO(cjihrig): Implement a TAP parser to read the child's stdout
+    // instead of just displaying it all if the child fails.
+    let err;
+
+    child.on('error', (error) => {
+      err = error;
+    });
+
+    const { 0: { 0: code, 1: signal }, 1: stdout, 2: stderr } = await SafePromiseAll([
+      once(child, 'exit', { signal: t.signal }),
+      child.stdout.toArray({ signal: t.signal }),
+      child.stderr.toArray({ signal: t.signal }),
+    ]);
+
+    if (code !== 0 || signal !== null) {
+      if (!err) {
+        err = new ERR_TEST_FAILURE('test failed', kSubtestsFailed);
+        err.exitCode = code;
+        err.signal = signal;
+        err.stdout = ArrayPrototypeJoin(stdout, '');
+        err.stderr = ArrayPrototypeJoin(stderr, '');
+        // The stack will not be useful since the failures came from tests
+        // in a child process.
+        err.stack = undefined;
+      }
+
+      throw err;
+    }
+  });
+  return subtest.start();
+}
+
+function run(options) {
+  if (options === null || typeof options !== 'object') {
+    options = kEmptyObject;
+  }
+  const { concurrency, timeout, signal, files } = options;
+
+  if (files !== undefined) {
+    validateArray(files, 'options.files');
+  }
+
+  const root = createTestTree({ concurrency, timeout, signal });
+  const testFiles = files ?? createTestFileList();
+
+  SafePromiseAll(ArrayPrototypeMap(testFiles, (path) => runTestFile(path, root)))
+    .then(() => root.postRun());
```

Comment:
this is still not resolved, but here's an alternative to https://github.com/nodejs/node/pull/44241/files/e5f6b85c3dafafb667c3e79d15410994d67c1203#r951907517
```suggestion
  await SafePromiseAll(ArrayPrototypeMap(testFiles, (path) => runTestFile(path, root)));
  root.postRun();
```

Reply:
the function is synchronous

Reply:
ah ok, in that case https://github.com/nodejs/node/pull/44241/files/e5f6b85c3dafafb667c3e79d15410994d67c1203#r951907517 (using `PromisePrototypeThen`) is still required.

Reply:
yes, I already implemented that

Reply:
yay, thanks! maybe i missed it in the confusing github review diffs :-)